pta-intelligent-search-harvester/src/main/java/fi/maanmittauslaitos/pta/search/HarvesterConfig.java

```diff
@@ -1,294 +1,369 @@
+			@SuppressWarnings("unchecked")
+			List<Double> coords = (List<Double>) original;
+			if (coords == null) {
+				return null;
+			} else {
+				return (coords.get(2) - coords.get(0)) * (coords.get(3) - coords.get(1));
 			}
 		});
 		bboxAreaFec.setField("geographicBoundingBoxArea");
 		configuration.getFieldExtractors().add(bboxAreaFec);
-		
+
+		// Best matching regions
+		configuration.getFieldExtractors().add(getBestMatchingRegions(bboxFec));
+
+
 		return factory.getDocumentProcessorFactory().createProcessor(configuration);
-		
+	}
+
+	private XPathFieldExtractorConfiguration getBestMatchingRegions(XPathFieldExtractorConfiguration bboxFec) {
+		ObjectMapper objectMapper = new ObjectMapper();
+		ObjectReader listReader = objectMapper.readerFor(new TypeReference<List<Double>>() {
+		});
+		Map<String, Region> countries = Region.readRegionResource(objectMapper, listReader, "data/well_known_location_bboxes_countries.json");
+		Map<String, Region> regions = Region.readRegionResource(objectMapper, listReader, "data/well_known_location_bboxes_regions.json");
+		Map<String, Region> subregions = Region.readRegionResource(objectMapper, listReader, "data/well_known_location_bboxes_subregions.json");
+		Map<String, Region> municipalities = Region.readRegionResource(objectMapper, listReader, "data/well_known_location_bboxes_municipalities.json");
+
+		XPathFieldExtractorConfiguration regionFec = (XPathFieldExtractorConfiguration) bboxFec.copy();
+		final GeographicBoundingBoxXPathCustomExtractor originalBboxCustomExtractor = (GeographicBoundingBoxXPathCustomExtractor) regionFec.getCustomExtractor();
+
+
+		BiFunction<Map<String, Region>, Region, Region.RegionScore> getBestRegionScore = (Map<String, Region> regionType, Region dataRegion) -> regionType.entrySet().stream()//
+				.filter(nameRegionEntry -> nameRegionEntry.getValue().intersects(dataRegion))//
+				.max(Comparator.comparing(entry -> entry.getValue().getIntersection(dataRegion)))//
+				.map(entry -> Region.RegionScore.create(entry.getKey(), entry.getValue().getIntersectionDividedByArea(dataRegion)))
+				.orElse(Region.RegionScore.createEmpty());
+
+
+		regionFec.setCustomExtractor((xPath, node) -> {
+			Object original = originalBboxCustomExtractor.process(xPath, node);
+			@SuppressWarnings("unchecked")
+			List<Double> coordinates = (List<Double>) original;
+			if (coordinates.isEmpty()) {
+				return objectMapper.createObjectNode();
+			}
+			Region dataRegion = new Region(coordinates);
+			Region.RegionScore country = getBestRegionScore.apply(countries, dataRegion);
+			Region.RegionScore region = getBestRegionScore.apply(regions, dataRegion);
+			Region.RegionScore subregion = getBestRegionScore.apply(subregions, dataRegion);
+			Region.RegionScore municipality = getBestRegionScore.apply(municipalities, dataRegion);
+
+			String template = "{\n" +
+					"  \"country\": {\n" +
+					"    \"location_name\": \"%s\",\n" +
```

Comment:
Once the field names are in Constants, the constants should be used here as well